### PR TITLE
team manager should get cancelled tasks

### LINKED ISSUE #4069: Team managers should be able to get tasks twice
The rule that a user can’t get a task again, which he already had, but which was reset and cancelled, should not be active for team managers and admins.
https://discuss.webknossos.org/t/reset-and-cancel/1458/3

### PR DESCRIPTION
**Pull request by youri-k**
- team manager should be able to get a task which they previously cancelled
- see discuss in issue

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- request task as a team manager
- cancel that task
- request another task
- get the same task again

### Issues:
- fixes #4069 

------
- [x] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)
- [x] Ready for review

### CODE REVIEW

File: app/models/task/Task.scala [hunk 1]
```diff
@@ -186,10 +186,10 @@ class TaskDAO @Inject()(sqlClient: SQLClient, projectDAO: ProjectDAO)(implicit e
...
              join webknossos.projects_ on webknossos.tasks_._project = webknossos.projects_._id
-             left join (select _task from webknossos.annotations_ where _user = '${userId.id}' and typ = '${AnnotationType.Task}') as userAnnotations ON webknossos.tasks_._id = userAnnotations._task
+             left join (select _task, state from webknossos.annotations_ where _user = '${userId.id}' and typ = '${AnnotationType.Task}') as userAnnotations ON webknossos.tasks_._id = userAnnotations._task
            where webknossos.tasks_.openInstances > 0
                  and webknossos.projects_._team in ${writeStructTupleWithQuotes(teamIds.map(t => sanitize(t.id)))}
-                 and userAnnotations._task is null
+                 and (userAnnotations._task is null or (((select isAdmin from webknossos.users_ where _id = '${userId.id}') or webknossos.projects_._team in (select _team from webknossos.user_team_roles where isTeamManager and _user = '${userId.id}')) and userAnnotations.state = '${AnnotationState.Cancelled}'))
```
Comment (jstriebel) [inline app/models/task/Task.scala]: You can't reuse `deleteAccessQ` easily here, can you?

Comment (youri-k) [reply app/models/task/Task.scala]: Sadly I don't think so because the attributes name are different after the `join`

**Comment by philippotto** [comment]
@youri-k What's the status here? Do you know how to tackle the problem @fm3 mentioned or do you need support? :)

**Comment by youri-k** [comment]
@philippotto I identified the cause of the bug and already tried fixing it. Unfortunately both of my two ideas didn't work out. This is on top of my priority list for tomorrow but if it's more urgent @fm3 might solve this faster.

**Comment by philippotto** [comment]
> @philippotto I identified the cause of the bug and already tried fixing it. Unfortunately both of my two ideas didn't work out. This is on top of my priority list for tomorrow but if it's more urgent @fm3 might solve this faster.

Cool :+1: Tomorrow sounds good! The lab got a workaround for the issue :)

**Comment by fm3** [comment]
@youri-k thanks for looking into this! A thought: If expressing this in SQL becomes too complex (and I also couldn’t quickly solve this) it might be a good time to further compose the query using scala. Like, check in advance if the requesting user is a team-manager/admin, and in this case insert further conditions into the query. This might also be more readable code in the end, compared to really weird join/subquery combinations :) What do you think?

**Comment by youri-k** [comment]
That sounds a whole lot easier and I also think it would make the code more readable. So I'll try that :slightly_smiling_face:

**Comment by youri-k** [comment]
@fm3 I think you can have a look again. I think the current solution works pretty well. The only thing that bothers me, is that team manager of one team can get cancelled tasks for another team. I don't know how big of a deal this is

**Comment by philippotto** [comment]
> I’d say the cross-team thing is fine for now. Seems like a rather unlikely scenario, let’s handle this when it comes up.

Just to clarify: As soon as there are two different teams (or organizations?) and cancelled tasks, a manager of team A can get tasks of team B? Nor sure whether we want to wait until this comes up :thinking: Will they be able to trace the task even though they don't belong to the right team?

**Comment by fm3** [comment]
No, not that bad, I think. 
Only if they also belong to that task’s team (as a user) and have a cancelled instance of that task, they can get it again. Right @youri-k ?

**Comment by youri-k** [comment]
Yes exactly @fm3

**Comment by philippotto** [comment]
Ok, I got it now (I think :D). Even with the low probability of this happening, the potential impact would be quite low from my understanding.

**Pull request merged**